gtsam_unstable/geometry/Similarity3.cpp

```diff
@@ -85,10 +85,110 @@ Point3 Similarity3::transformFrom(const Point3& p, //
   return s_ * q;
 }
 
+Pose3 Similarity3::transformFrom(const Pose3& T) const {
+  Rot3 R = R_.compose(T.rotation());
+  Point3 t = Point3(s_ * (R_.matrix() * T.translation().vector() + t_.vector()));
+  return Pose3(R, t);
+}
+
 Point3 Similarity3::operator*(const Point3& p) const {
   return transformFrom(p);
 }
 
+// Refer to: http://www5.informatik.uni-erlangen.de/Forschung/Publikationen/2005/Zinsser05-PSR.pdf Chapter 3
+Similarity3 Similarity3::Align(const std::vector<Point3Pair>& abPointPairs) {
+  const size_t n = abPointPairs.size();
+  if (n < 3) throw std::runtime_error("input should have at least 3 pairs of points");  // we need at least three pairs
+
+  // calculate centroids
+  Point3 aCentroid(0, 0, 0), bCentroid(0, 0, 0);
+  for (const Point3Pair& abPair : abPointPairs) {
+    aCentroid += abPair.first;
+    bCentroid += abPair.second;
+  }
+  double f = 1.0 / n;
+  aCentroid *= f;
+  bCentroid *= f;
+
+  // Add to form H matrix
+  Matrix3 H = Z_3x3;
+  vector<Point3Pair> d_abPairs;
+  for (const Point3Pair& abPair : abPointPairs) {
+    Point3 da = abPair.first - aCentroid;
+    Point3 db = abPair.second - bCentroid;
+    d_abPairs.push_back(make_pair(da,db));
+    H += da * db.transpose();
+  }
+
+  // ClosestTo finds rotation matrix closest to H in Frobenius sense
+  Rot3 aRb = Rot3::ClosestTo(H);
+
+  // Calculate scale
+  double x = 0;
+  double y = 0;
+  for (const Point3Pair& d_abPair : d_abPairs) {
+    Point3 da = d_abPair.first;
+    Point3 db = d_abPair.second;
+    Vector3 Rdb = aRb * db;
+    y += da.transpose() * Rdb;
+    x += Rdb.transpose() * Rdb;
+  }
+  double s = y / x;
+
+  Point3 aTb = (Point3(aCentroid) - s * (aRb * Point3(bCentroid))) / s;
+  return Similarity3(aRb, aTb, s);
+}
+
+//  Use the geodesic L2 mean to solve the mean of rotations,
+//  Refer to: http://users.cecs.anu.edu.au/~hongdong/rotationaveraging.pdf (on page 18)
+Rot3 Similarity3::rotationAveraging(const std::vector<Rot3>& rotations, double error) {
+  Rot3 R = rotations[0];
+  const size_t n = rotations.size();
+  Vector3 r;
+  r << 0, 0, 0;
+  while (1) {
+    for (const Rot3 R_i : rotations) {
+      r += Rot3::Logmap(R.inverse().compose(R_i));
```

Comment:
I'm a bit confused about this line. For the first iteration, we would get `R₁.inverse().compose(R₁) = I`. I am guessing this is correct since we assume the first pose is "start"?

Please confirm.